File: src/main/java/com/zhan_dui/viewmodel/MemoViewModel.java
package com.zhan_dui.viewmodel;

import android.app.Application;
import android.content.ContentUris;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.zhan_dui.data.Memo;
import com.zhan_dui.data.MemoProvider;
import com.zhan_dui.repository.RoomMemoRepository;

import java.util.List;

/**
 * ViewModel for memo-related UI data.
 */
public class MemoViewModel extends AndroidViewModel {
    private final RoomMemoRepository memoRepository;
    private final LiveData<List<Memo>> allMemos;
    private final MutableLiveData<Memo> selectedMemo = new MutableLiveData<>();
    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();

    public MemoViewModel(@NonNull Application application) {
        super(application);
        memoRepository = new RoomMemoRepository(application);
        allMemos = memoRepository.getAllMemos();
    }

    public LiveData<List<Memo>> getAllMemos() {
        return allMemos;
    }

    public LiveData<Memo> getSelectedMemo() {
        return selectedMemo;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    /**
     * Select a memo for editing/viewing
     */
    public void selectMemo(Memo memo) {
        selectedMemo.setValue(memo);
    }

    /**
     * Clear selected memo
     */
    public void clearSelectedMemo() {
        selectedMemo.setValue(null);
    }

    /**
     * Create a new memo
     */
    public Uri createMemo(String content) {
        if (content == null || content.trim().isEmpty()) {
            errorMessage.setValue("Memo content cannot be empty");
            return null;
        }

        Memo memo = new Memo();
        memo.setContent(content);
        long id = memoRepository.insertMemoSync(memo);
        if (id > 0) {
            // Return a ContentProvider-style URI for compatibility
            return ContentUris.withAppendedId(MemoProvider.MEMO_URI, id);
        } else {
            errorMessage.setValue("Failed to insert memo");
            return null;
        }
    }

    /**
     * Update an existing memo
     */
    public boolean updateMemo(Memo memo) {
        if (memo == null) {
            errorMessage.setValue("Memo cannot be null");
            return false;
        }

        if (memo.getContent() == null || memo.getContent().trim().isEmpty()) {
            // Empty content means delete
            return deleteMemo(memo.getId()) > 0;
        }

        memoRepository.updateMemo(memo, new RoomMemoRepository.UpdateCallback() {
            @Override
            public void onUpdated(boolean success) {
                if (!success) {
                    errorMessage.setValue("Failed to update memo");
                }
            }

            @Override
            public void onError(Exception e) {
                errorMessage.setValue("Failed to update memo: " + e.getMessage());
            }
        });
        // Return true immediately assuming operation will succeed
        return true;
    }

    /**
     * Delete a memo
     */
    public int deleteMemo(int id) {
        memoRepository.deleteMemo(id, new RoomMemoRepository.DeleteCallback() {
            @Override
            public void onDeleted(boolean success) {
                if (!success) {
                    errorMessage.setValue("Failed to delete memo");
                }
            }

            @Override
            public void onError(Exception e) {
                errorMessage.setValue("Failed to delete memo: " + e.getMessage());
            }
        });
        // Return 1 immediately assuming operation will succeed
        return 1;
    }

    /**
     * Refresh memos from repository
     */
    public void refreshMemos() {
        memoRepository.refreshMemos();
    }

    /**
     * Get memo by ID
     */
    public Memo getMemoById(int id) {
        return memoRepository.getMemoById(id);
    }
}

File: app/src/main/java/com/zhan_dui/viewmodel/MemoViewModel.kt
package com.zhan_dui.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.zhan_dui.data.MemoEntity
import com.zhan_dui.repository.MemoRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlin.coroutines.Continuation

/**
 * ViewModel for memo-related UI data.
 */
open class MemoViewModel(application: Application) : AndroidViewModel(application) {
    private val memoRepository: MemoRepository = MemoRepository(application)
//    private val allMemos: LiveData<List<MemoEntity>>
    private val selectedMemo: MutableLiveData<MemoEntity?> = MutableLiveData<MemoEntity?>()
    private val errorMessage = MutableLiveData<String?>()

//    init {
//        allMemos = memoRepository.getAllMemos()
//    }

    fun getAllMemos(): LiveData<List<MemoEntity>> {
        return memoRepository.getAllMemos()
    }

    fun getSelectedMemo(): LiveData<MemoEntity?> {
        return selectedMemo
    }

    fun getErrorMessage(): LiveData<String?> {
        return errorMessage
    }

    /**
     * Select a memo for editing/viewing
     */
    fun selectMemo(memo: MemoEntity?) {
        selectedMemo.value = memo
    }

    /**
     * Clear selected memo
     */
    fun clearSelectedMemo() {
        selectedMemo.value = null
    }

    /**
     * Create a new memo
     */
    fun createMemo(memo: MemoEntity): Long {
        if (memo.content == null || memo.content?.trim { it <= ' ' }?.isEmpty() == true) {
            errorMessage.value = "Memo content cannot be empty"
            return -1L
        }

        return try {
            val id = memoRepository.insertMemo(memo)
            if (id <= 0) {
                errorMessage.postValue("Failed to insert memo")
            }
            id
        } catch (e: Exception) {
            errorMessage.postValue("Error: " + e.message)
            -1L
        }
    }

    /**
     * Update an existing memo
     */
    fun updateMemo(memo: MemoEntity?) {
        if (memo == null) {
            errorMessage.setValue("Memo cannot be null")
            return
        }

        if (memo.content == null || memo.content?.trim { it <= ' ' }?.isEmpty() == true) {
            // Empty content means delete
            deleteMemo(memo)
            return
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val rows = memoRepository.updateMemo(memo)
                if (rows <= 0) {
                    errorMessage.postValue("Failed to update memo")
                }
            } catch (e: Exception) {
                errorMessage.postValue("Error: " + e.message)
            }
            null
        }
    }

    /**
     * Delete a memo
     */
    fun deleteMemo(memo: MemoEntity) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val rows = memoRepository.deleteMemo(memo)
                if (rows <= 0) {
                    errorMessage.postValue("Failed to delete memo")
                }
            } catch (e: Exception) {
                errorMessage.postValue("Error: " + e.message)
            }
            null
        }
    }

    /**
     * Refresh memos from repository
     */
    fun refreshMemos() {
        memoRepository.refreshMemos()
    }

    /**
     * Get memo by ID
     */
    fun getMemoById(id: Long): MemoEntity? {
        return memoRepository.getMemoById(id)
    }
}